presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayNotContainsToAntiJoin.java

```diff
@@ -0,0 +1,223 @@
+
+        RowExpression arrayContainsExpression = arrayNotContainsExpression.getChildren().get(0);
+        RowExpression array = ((CallExpression) arrayContainsExpression).getArguments().get(0);
+        RowExpression element = ((CallExpression) arrayContainsExpression).getArguments().get(1);
+        boolean arrayAtLeftInput = leftColumns.containsAll(VariablesExtractor.extractAll(array));
+        PlanNode inputWithArray = arrayAtLeftInput ? joinNode.getLeft() : joinNode.getRight();
+
+        if (!isFromScalarSubquery(context, inputWithArray)) {
+            // rewrite is incorrect if array input has more than 1 row or columns included in the output: if the source of CROSS JOIN was a subquery these conditions are guaranteed
+            System.out.println("Not from scalar subquery");
+            return Result.empty();
+        }
+        final Type type = element.getType();
+        CallExpression arrayDistinct = call(functionAndTypeManager, "array_distinct", new ArrayType(type),
+                call(functionAndTypeManager, "remove_nulls", new ArrayType(type), array));
+        VariableReferenceExpression arrayDistinctVariable = context.getVariableAllocator().newVariable(arrayDistinct);
+        PlanNode project = PlannerUtils.addProjections(inputWithArray, context.getIdAllocator(), ImmutableMap.of(arrayDistinctVariable, arrayDistinct));
+        VariableReferenceExpression unnestVariable = context.getVariableAllocator().newVariable("field", type);
+        UnnestNode unnest = new UnnestNode(inputWithArray.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                project,
+                project.getOutputVariables(),
+                ImmutableMap.of(arrayDistinctVariable, ImmutableList.of(unnestVariable)),
+                Optional.empty());
+
+        PlanNode newLeftNode;
+
+        if (arrayAtLeftInput) {
+            newLeftNode = joinNode.getRight();
+        }
+        else {
+            newLeftNode = joinNode.getLeft();
+        }
+
+        // if element is not a VariableReferenceExpression, push the expression into a Project node so the variable can be used in equijoins
+        VariableReferenceExpression keyVariable;
+        if (element instanceof VariableReferenceExpression) {
+            keyVariable = (VariableReferenceExpression) element;
+        }
+        else {
+            keyVariable = context.getVariableAllocator().newVariable(element);
+            newLeftNode = addProjections(newLeftNode, context.getIdAllocator(), ImmutableMap.of(keyVariable, element));
+        }
+
+        JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(keyVariable, unnestVariable);
+
+        List<VariableReferenceExpression> newOutputColumns = Stream.concat(newLeftNode.getOutputVariables().stream(), unnest.getOutputVariables().stream()).collect(Collectors.toList());
```

Comment:
This assumes that no output from array input side, should we add such check at the beginning?

Reply:
the cross join does not do any projection pruning so the output will always be the union of the outputs of the children. Here we just make sure we project out any variables introduced by the unnest operator

Reply:
I mean for this rewrite, it's not a equivalent rewrite if there are output from the array side? As the left join will only output when there is no match, and all fields from the right side will be NULL, hence not equivalent to the inner join. The rewrite here should only apply to the case where there is no output from the array side, and we need to add the constraint in the query pattern.

Reply:
Indeed, this is why the transformation only kicks in if the array side is coming from a scalar subquery (which have EnforceSingleRowNode on top). A cleaner way would be to pass required columns from top nodes as part of the context so that we can have that check in the rule, but we currently don't have that capability.
I also have tests to make sure we don't produce wrong results if some column from the array side is projected

Reply:
Oh, I see. The challenge is that there is no easy way to get the projected output for the join. Then it's fine as we only output the input from the left side in the left join.